New docker image for Go+Docker+Heroku

philpennock opened the pull request:
We want the Docker tools, to talk to a remote Docker server, for use
with Circle CI's `setup_remote_docker` so that we can build Docker
images when running in Docker.  We also want Go, but now can just use
upstream Golang images, and we add Heroku as before, and our packages.

This image drops the user to a non-root runtime user, with no way back
to root.  Not a problem: if we need more privs, we can rebuild the
container.

philpennock committed:
New docker image for Go+Docker+Heroku

We want the Docker tools, to talk to a remote Docker server, for use
with Circle CI's `setup_remote_docker` so that we can build Docker
images when running in Docker.  We also want Go, but now can just use
upstream Golang images, and we add Heroku as before, and our packages.

This image drops the user to a non-root runtime user, with no way back
to root.  Not a problem: if we need more privs, we can rebuild the
container.

cee-dub commented on docker-go/Dockerfile:
```diff
@@ -0,0 +1,64 @@
+# Want to just remove the Go image's GOPATH setting but while we can replace
+# ENV in Docker, we can't unset.  So suck it up and just set it, old style.
+ENV GOPATH=/home/${RUNTIME_USER}/go
+ENV PATH ${GOPATH}/bin:/usr/local/go/bin:$PATH
```
Missing equals sign?

philpennock replied:
Nope, although for consistency I should use one or the other.
https://docs.docker.com/engine/reference/builder/#env

cee-dub commented:
In general, this looks promising. Am.i correct that we'd use this image in CircleCI to run tests and to run the build of a different Docker image  that replaces slugs for production deployment?

philpennock commented:
You are correct.  We'd replace the image we're using now, from the `circlego` sub-directory.

We basically need the Docker binary client available, on top of other stuff we have already added.  It looked like using the official golang images presented a marginally saner base than Circle's Golang one.

philpennock committed:
Be consistent in use of = for ENV directives

Without a specific need to use different parsing rules, we should be
consistent.  So use `K=V` (allowing `K="V 2"` form) by preference.

cee-dub commented on docker-go/Dockerfile:
```diff
@@ -0,0 +1,64 @@
+
+WORKDIR /home/${RUNTIME_USER}
+# We don't use /go because we don't build as root and 777 permissions are daft
+RUN rm -rf /go
```
Where does the go runtime live in this image? Obviously not in here.

philpennock replied:
In `/usr/local/go`

The start of the file:
```
ARG GOLANG_BASE_IMAGE=1.9.3-stretch
FROM golang:${GOLANG_BASE_IMAGE}
```

So:

```
% docker run -it --rm golang:1.9.3-stretch
root@7aaa093586e5:/go# ls /go
bin  src
root@7aaa093586e5:/go# ls /usr/local/go
AUTHORS		 CONTRIBUTORS  PATENTS	  VERSION  bin	 doc	      lib   pkg		src
CONTRIBUTING.md  LICENSE       README.md  api	   blog  favicon.ico  misc  robots.txt	test
root@7aaa093586e5:/go#
```

cee-dub commented on docker-go/Dockerfile:
```diff
@@ -0,0 +1,64 @@
+RUN go version && \
+	go get -u -v github.com/jstemmer/go-junit-report && \
+	go get -u -v github.com/nbio/slugger && \
+	go get -u -v github.com/nbio/cart && \
```
How do we keep these tools up to date in the image? Is doing this sufficient to update them each time the Dockerfile is executed?

philpennock replied:
Yes: every time Dockerfile is run, it's effectively run in a clean filesystem.  The `-u` flag to `go get` does an Update.  So building the image will update these tools.

philpennock committed:
Go 1.9.4, add 'dep', remove other Docker files

cee-dub commented on README.md:
```diff
@@ -1,2 +1,8 @@
+
 CI support for Domainr
+
+Used to create: `domainr/docker-go:go1.9.4-docker17.12.0-ce` or the like
```
This name is configured in Docker hub, right? The config there will build `Dockerfile` at the root as `:latest`.

philpennock replied:
Yes, that was "what I was doing locally" and in fact this statement is wrong.

cee-dub commented on README.md:
```diff
@@ -1,2 +1,8 @@
+Used to create: `domainr/docker-go:go1.9.4-docker17.12.0-ce` or the like
+
+Possibly used in future builds, having the Docker client tool to talk to a
+remote docker server, so should work with Circle CI's `setup_remote_docker`.
```
It's going to be used immediately. :D

ydnar replied:
Yep

cee-dub reviewed (approved):
Let's see if it blends!

ydnar commented on README.md:
```diff
@@ -1,2 +1,8 @@
+
 CI support for Domainr
+
+Used to create: `domainr/docker-go:go1.9.4-docker17.12.0-ce` or the like
```
Can we call this `domainr/ci`?

ydnar reviewed (approved):
LGTM, can we call this `domainr/ci` instead of the longer mangled image name? Does that matter?

philpennock commented:
It's called whatever you tag it; auto-builds will tag it as `domainr/ci:latest` probably, depending upon what Eric set up.  That text refers to what I was doing before, missed another change.

philpennock committed:
README fixes

cee-dub commented:
Thumbs up!

philpennock committed:
Add zip package
